compiler/rustc_mir_build/src/thir/pattern/deconstruct_pat.rs

```diff
@@ -1343,50 +1348,70 @@ impl<'p, 'tcx> DeconstructedPat<'p, 'tcx> {
             }
             PatKind::Constant { value } => {
-                if let Some(int_range) = IntRange::from_constant(cx.tcx, cx.param_env, *value) {
-                    ctor = IntRange(int_range);
-                    fields = Fields::empty();
-                } else {
-                    match pat.ty.kind() {
-                        ty::Float(_) => {
-                            ctor = FloatRange(*value, *value, RangeEnd::Included);
-                            fields = Fields::empty();
-                        }
-                        ty::Ref(_, t, _) if t.is_str() => {
-                            // We want a `&str` constant to behave like a `Deref` pattern, to be compatible
-                            // with other `Deref` patterns. This could have been done in `const_to_pat`,
-                            // but that causes issues with the rest of the matching code.
-                            // So here, the constructor for a `"foo"` pattern is `&` (represented by
-                            // `Single`), and has one field. That field has constructor `Str(value)` and no
-                            // fields.
-                            // Note: `t` is `str`, not `&str`.
-                            let subpattern =
-                                DeconstructedPat::new(Str(*value), Fields::empty(), *t, pat.span);
-                            ctor = Single;
-                            fields = Fields::singleton(cx, subpattern)
-                        }
-                        // All constants that can be structurally matched have already been expanded
-                        // into the corresponding `Pat`s by `const_to_pat`. Constants that remain are
-                        // opaque.
-                        _ => {
-                            ctor = Opaque;
-                            fields = Fields::empty();
-                        }
+                match pat.ty.kind() {
+                    ty::Bool | ty::Char | ty::Int(_) | ty::Uint(_) | ty::Float(_) => {
+                        use rustc_apfloat::Float;
+                        ctor = match fast_try_eval_bits(cx.tcx, cx.param_env, value) {
+                            Some(bits) => match pat.ty.kind() {
+                                ty::Bool | ty::Char | ty::Int(_) | ty::Uint(_) => {
+                                    IntRange(IntRange::from_bits(cx.tcx, pat.ty, bits))
+                                }
+                                ty::Float(ty::FloatTy::F32) => {
+                                    let value = rustc_apfloat::ieee::Single::from_bits(bits);
+                                    F32Range(value, value, RangeEnd::Included)
+                                }
+                                ty::Float(ty::FloatTy::F64) => {
+                                    let value = rustc_apfloat::ieee::Double::from_bits(bits);
+                                    F64Range(value, value, RangeEnd::Included)
+                                }
```

Comment:
Readability-wise, should we move those 2 cases to the enclosing matches, duplicating the call to `fast_try_eval_bits`?